README.md
# Sensof
Si4010 based wireless sensors(temperature, humidity, etc.) firmware and utilities

# Sensof-HT
Sensof-HT is a temperature and humidity sensor that is build using a Si4010 MCU
and a HTU21 sensor.

## Circuit diagram
The Si4010 connects to a HTU21D, Si7020 or compatible sensor through I2C. The
I2C bus is connected to the Si4010 on the following pins:

   GPIO1 = SCL
   GPIO2 = SDA

Voltage level should be between 3.6 - 1.9 Volt. A 3V CR2032 works fine.

Some ready made Si4010 and HTU21D or Si7020 sensor are available on the
internet. These only need to be connected together and connected to a battery.

This repository also contains a custom Sensof-HT PCB. However this is currently
still under development. See hardware/README.md for more information.

## Receiver
I currently use a RFM22B module from modtronicsaustralia.com connected to a
Raspberry Pi. This is a Si4430 based transceiver module. For the source code of
the receiving program see the src/si443x_sensof directory.

# Sensof Frame Format
The Sensof frames are compatible with the Si4430/EZRadioPro receiver, and have
the following structure:

    +-----------+-----------+--------------+------------+
    | Preamble  | Sync Word | Enc. Payload | CRC-16-IBM |
    | 13 * 0xaa | 0x2d 0xd4 |   16-byte    |   2-byte   |
    +-----------+-----------+--------------+------------+

The CRC-16 uses the 0x8005 polynomial. The payload is a single AES-128
encrypted block, see [Encrypted Payload].

## Encrypted Payload
The payload is encrypted using AES-128 in ECB mode. To make this secure
against various attacks some mandatory field within the encrypted block have
been defined. The payload is formatted as follows:

    +---------+----------+------------------+
    | Dev. ID | Frame ID | Application data |
    | 4-byte  |  4-byte  |      8-byte      |
    +---------+----------+------------------+

The device ID identifies the source of the message. The Si4010 32-bit uniq
chip ID shall be used for this. The receiver will ignore frames that are from
an unknown source.

The 32-bit frame ID uniquely identifies the frame and allows duplicate frame
detection. The frame ID will be incremented sequentially, starting at 0. The
receiver must only accept frames with a frame ID above the last received frame
ID, and which is a 'reasonable amount higher' than the previous frame ID. See
the [Sensof Security Rational] chapter to get an indication of what a
'reasonable amount higher' means.

The remaining 8 byte can be used for application specific data. It is not
possible to have more or less then 8-byte payload in a single frame. If less
bytes are required the remaining bytes must be set to a known default. For
security reasons the receiver must verify these bytes. If more then 8 bytes
payload are required the payload must be split across multiple frames.

A full explanation of the security model is described below in [Sensof Security
Rational].

## Sensof-HT Frame Payload
The Sensof-HT frames contain the following payload:

 - Temperature(uint16_t): Temperature value as returned by HTU21 IC
 - Humidity(uint16_t): Humidity value as returned by HTU21 IC
 - Padding(uint32_t): '0' padding

See the HTU21 datasheet for the formulas to convert the temperature and
humidity to human readable values.

## SI4430 Compatibility
To receive the frames with the Si4430 receiver the following settings need to
be changed from the default configuration:

 - Header Length = 0
 - Fixed packet length = True
 - Packet length = 16

# Sensof Security Rational
The payload of all Sensof frames are encrypted to guarantee confidentiality and
integrity.

## Confidentiality
The AES algorithm is used with a 128-bit key. Since only a single block is
encrypted the strength of the encryption will be equal to that of the AES
algorithm.

All nodes in a system use the same pre-shared encryption key. This means that
if a single node in the system is suspected to be compromised the whole system
needs to be re-keyed.

## Integrity
The system depends on the avalanche effect of the AES block cipher protect
it integrity.

This means that if an adversary changes a single bit in the cipher text, that
the plain text changes completely and in an to the adversary unpredictable
manner.

By embedding known values into the encrypted blocks the receiver can verify
the decrypted data by checking if the known values are correct. In case of the
Sensof frames the device ID and frame ID are used for this. The change that an
adversary generates a frame that is accepted by the receiver would (probably)
be (somewhere in the order of):

    P = (N / 2^32) * (W / 2^32)

Where:
 N = Number of known nodes
 W = Size of accepted frame ID window

Of course any structure in the remaining 8 byte of payload will contribute to
this integrity protection. Therefor it is suggested to set any unused bytes to
'0', or any other fixed value, and verify these bytes on the receiver.

### Replay attack
To protect against replay attacks every frame has a 32-bit counter embedded.
The receiver must only accept frames with a frame ID above the last received
frame ID, and which is not more then a reasonable amount higher than the
previous frame ID.

In this case reasonable is the maximum amount of packets that one can expect to
be lost during normal operation.

It is essential that the receiver is initialized with the correct replay
counter at startup.  Just accepting the replay counter of the first received
frame as the current replay counter is not suggested. This allows an adversary
to replay a previously recorded stream of frames.

A suggested way is too let the receiver (periodically) store the current replay
counters per device and time to non-volatile storage. Upon restart of the
receiver these values should be read from storage and based on this the last
frame ID per device should be determined.
 In case of senders that send with a fix periodic interval this calculation can
be pretty accurate. For non-periodic senders other measures need to be taken
depending on the required security level.

If a sender losses its frame ID there are multiple possibilities:

 1. Reload node with expected frame ID based on the receiver's state
 2. Change the Device ID of the node
 3. Change the encryption key for the whole system

### Delay attack
In a delay attack the adversary provides the receiver with outdated
information. For example in a burglar alarm delaying the measurements of the
sensors for an hour would allow an adversary to break into the building with
the alarm only triggering after an hour.

To preform this attack the adversary would need to be able to prevent frames of
a node to reach the receiver, and record and replay those frames. This would
work if the receiver only relies on the frame ID from the frames to determine
the expected frame ID.

To prevent this attack for nodes that send packets with a fixed period the
receiver can protect against this by taking into account the frame interval,
current time and time of the last received frame to calculate what the expected
frame ID should be. Of course this system must leave some room for in accuracy
in the sending interval. But in general this should greatly reduce the attack
surface(Sorry, not going to prove statement this here...).

For nodes that do not have a periodic interval of sending frames the current
system does **NOT** offer any protection. If this is required an alternative
system where a time value is used instead of a frame ID can be considered.


hardware/CHANGES.md
# 0002-00: Sensof HT v1 PCB
## 0002-00-00
 * Initial PCB

## 0002-00-01
 * Change GND fill zone around matching network to be closer to SiLabs reference design
 * Added 'SN' field in silk screen for writing down serial number
 * Upgraded to KiCad 5.0
 * Switched to di_kicad_library repository for symbols & footprints


# 0002-01: Top layer solder stencil
## 0002-01-00
 * Initial stencil for 0002-00-00 PCB


# 0003-00: Sensof HT v2 PCB
## 0003-00-00
 * Initial PCB

## 0003-00-01
 * Upgraded to KiCad 5.0
 * Switched to di_kicad_library repository for symbols & footprints
 * Change test pad positions


# 0003-01: Top layer solder stencil
## 0003-01-00
 * Initial top stencil for 0003-00-00 PCB


# 0003-02: Bottom layer solder stencil
## 0003-02-00
 * Initial bottom stencil for 0003-00-00 PCB


firmware/src/rf.c
#include "rf.h"

#include <stdlib.h>
#include <string.h>

#include "si4010.h"
#include "si4010_api_rom.h"
#include "crc.h"

#ifndef BOARD_VERSION
#define BOARD_VERSION 2
#endif

// Transmission parameters

// Values are calculated with si4010_calc_regs_110107.ods. Used input data:
// Package: v1: 10-pin MSOP; v2: 14-pin SOIC
//
// Antenna:
// - Alpha: 0 (we don't mind radiant power fluctuation over temperature)
// - Approx. Eff:
//   - v1: 35 % (from ACAG1204-433-T-1314083.pdf datasheet) 
//   - v2: 12.5 % (default)
// - Reactance change: 15 % (default)
// - Manual Impedance: no
//
// Power Amplifier Setup:
// - Total PA Power: 10.5 dBm (Above this sheet starts to complain)
// - Max Radiant Power
// - Center Freq: 433.92 MHz (Center of 433/434 ISM band)
// - Nom. Cap Word: 
//   - v1: 23 (from AN369: SI4010 Antenna Interface and Matching Network Guide")
//   - v2: 80 (from AN369: SI4010 Antenna Interface and Matching Network Guide")
// - External diff Cap: 0
//
// Serializer Setup:
// - Data rate: 9.6 kbits/s
// - Encoding: NRZ Only (eg. no encoding)
// - Modulation: FSK
// - FSK Deviation: 59 kHz
//
// Calculated Values:
// - Actual symbol rate: 9.59 kbit/s
// - Expected deviation: 59.05 kHz

const tPa_Setup code rPaSetup = {
#if BOARD_VERSION==1 // Sensof v1 - Ceramic antenna
	/*.fAlpha      =*/ 0,
	/*.fBeta       =*/ 0.2455, // Calculated: 0.2455 (MSOP-10), 0.2499 (SOIC-14)
	/*.bLevel      =*/ 76, // Max according to sheet is: 76 @ max. drv.
	/*.bMaxDrv     =*/ 1,
	/*.wNominalCap =*/ 23,
#elif BOARD_VERSION==2 // Sensof v2 - PCB antenna
	/*.fAlpha      =*/ 0,
	/*.fBeta       =*/ 0.1755, // Calculated: 0.1674 (MSOP-10), 0.1755 (SOIC-14)
	/*.bLevel      =*/ 76, // Max according to sheet is: 76 @ max. drv.
	/*.bMaxDrv     =*/ 1,
	/*.wNominalCap =*/ 80,
#else
#error "BOARD_VERSION incorrect or not defined"
#endif
};

/**
 * PaSetup structure pointer.
 *
 * Since code and xdata memory are the same it is possible to directly access 
 * constant structures in code memory. Normally the compiler would copy the 
 * structure from the code memory into xdata memory, causing two identical 
 * copies to exist.
 */
const tPa_Setup xdata  *prPaSetup = (const tPa_Setup xdata  *) &rPaSetup; 

const tOds_Setup code rOdsSetup = {
	/*.bModulationType =*/ 1,  // Use FSK
	/*.bClkDiv     =*/ 5,
	/*.bEdgeRate   =*/ 0,
	/*.bGroupWidth =*/ 7,
	/*.wBitRate    =*/ 417,   // Bit width in seconds = (ods_datarate*(ods_ck_div+1))/24MHz
	/*.bLcWarmInt  =*/ 0,
	/*.bDivWarmInt =*/ 5,
	/*.bPaWarmInt  =*/ 4,
};
/**
 * OdsSetup structure pointer.
 *
 * Since code and xdata memory are the same it is possible to directly access 
 * constant structures in code memory. Normally the compiler would copy the 
 * structure from the code memory into xdata memory, causing two identical 
 * copies to exist.
 */
const tOds_Setup xdata *prOdsSetup = (const tOds_Setup xdata *) &rOdsSetup;

const float fFreq = 433979050.0;
const float fFdiv = 102; // 59 KHz @ 433.92 MHz

// Frame data
/**
 * Frame data buffer
 *
 * Sensof frames as send through the air. These frames are compatible with
 * the Si4430/EZRadioPro receiver. The following settings differ from the 
 * default Si4430 configuration:
 *
 * - Header Length = 0
 * - Fixed packet length = True
 * - Packet length = 16
 */
const uint8_t code frame_data[] = {
	//reg 0x35[7:3] (number is amount of 4-bit nibbles!!!), reg 0x70[3]
	0xaa, 0xaa, 0xaa, 0xaa, 0xaa, 0xaa, 0xaa, 0xaa, // preamble
	0xaa, 0xaa, 0xaa, 0xaa, 0xaa, 			// preamble
	//reg 0x33[2:1], reg 0x36-0x39
	0x2d, 0xd4,                             // sync word 3 & 2
	//reg 0x33[6:4], reg 0x43-0x4a
	//0x11, 0x22,                             // Header 3 & 2
	//reg 0x33[3]
	//0x10,                                   // packet len
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Data
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Data
	//reg 0x30[5,2,1:0]
	0x00, 0x00                              // CRC-16-IBM
};
#define LEN_PREAMBLE	(13)
#define LEN_SYNC_WORD	(2)
#define	LEN_HEADER	(0)
#define	LEN_PLEN	(0)
#define LEN_PAYLOAD	(16)
#define LEN_CRC		(2)

#define OFFSET_SYNC_WORD (LEN_PREAMBLE)
#define OFFSET_HEADER	(OFFSET_SYNC_WORD + LEN_SYNC_WORD)
#define OFFSET_PLEN	(OFFSET_HEADER + LEN_HEADER)
#define OFFSET_PAYLOAD	(OFFSET_PLEN + LEN_PLEN)
#define OFFSET_CRC	(OFFSET_PAYLOAD + LEN_PAYLOAD)

#define LEN_FRAME	(OFFSET_CRC + LEN_CRC)

void rf_setup()
{

	// Set DMD interrupt to high priority,
	// any other interrupts have to stay low priority
	PDMD=1;

	// Call the system setup. This just for initialization.
	// Argument of 1 just configures the SYS module such that the
	// bandgap can be turned off if needed.  
	vSys_Setup( 1 );

	// Setup the bandgap for working with the temperature sensor here.
	// bSys_FirstBatteryInsertWaitTime set to non zero value. 
	vSys_BandGapLdo( 1 );

	// Required bPA_TRIM clearing before calling vPa_Setup() 
	vPa_Setup( prPaSetup );

	// ODS setup 
	vOds_Setup( prOdsSetup );

	// Setup the STL encoding for none.
	vStl_EncodeSetup( bEnc_NoneNrz_c, NULL );

	// Setup and run the frequency casting. 
	vFCast_Setup();

	// Disable Bandgap and LDO till needed
	vSys_BandGapLdo(0);
}

static void rf_transmit_frame(float freq, float fdiv, BYTE xdata *pbFrameHead, BYTE bLen, BYTE cnt)
{
	// Enable the Bandgap and LDO
	vSys_BandGapLdo(1);

	// Tune to the right frequency and set FSK ferquency adjust
	vFCast_Tune(freq);
	vFCast_FskAdj(fdiv);

	// Wait for a temperature sample and adjust oscilator with it
	while ( 0 == bDmdTs_GetSamplesTaken() ) {}
	vPa_Tune( iDmdTs_GetLatestTemp() );

	// Run a single TX loop 
	vStl_PreLoop();
	while (cnt != 0) {
		vStl_SingleTxLoop(pbFrameHead, bLen);
		cnt--;
	}
	vStl_PostLoop();

	// Disable Bandgap and LDO to save power
	vSys_BandGapLdo(0);
}

static uint8_t reverse_bits(uint8_t b)
{
	// reverse bit order
        b = ((b & 0xaa) >> 1) | ((b & 0x55) << 1);
        b = ((b & 0xcc) >> 2) | ((b & 0x33) << 2);
        b = ((b & 0xf0) >> 4) | ((b & 0x0f) << 4);

	return b;
}

void ezradiopro_send_frame(uint8_t *payload, uint8_t len)
{
	int i;

	uint16_t crc;
	uint8_t xdata *buf = (uint8_t xdata *) frame_data;

	if (len > 16) len = 16;
	memcpy(&buf[OFFSET_PAYLOAD], payload, len);

	// Calculate CRC-16
	crc = 0;
	for (i=OFFSET_HEADER; i < OFFSET_CRC; i++) {
		crc = crc16(crc, buf[i]);
	}
	buf[OFFSET_CRC] = (crc >> 8) & 0xff;
	buf[OFFSET_CRC + 1] = crc & 0xff;


#if 0 // only needed for encoding that changes frame length(ie. manchester)
	{
		BYTE xdata abFrameArray[LEN_FRAME * 2];
		uint8_t *pbFrameHead;           // Pointer to frame Head 

		for (i=0; i < LEN_FRAME; i++) {
			buf[i] ^= 0xff;
			pbFrameHead = encode_manchester(pbFrameHead, buf[i]);
		}

		rf_transmit_frame(fFreq, fFdiv, abFrameArray, pbFrameHead - abFrameArray, 1);
	}
#else
	for (i=0; i < LEN_FRAME; i++) {
		// Invert bits, by default Si4430 has different notion of '1' and '0'
		//TODO: can't we just fix this on receiver?
		buf[i] ^= 0xff;
		// Revert bits since Si4010 sends out LSB first
		buf[i] = reverse_bits(buf[i]);
	}


	// Transmit
	rf_transmit_frame(fFreq, fFdiv, buf, LEN_FRAME, 1);
#endif
}
